Document libhackrf functions using Doxygen

Sasszem opened the pull request:
I added doxygen documentations to the `hackrf.h` file, set up a `Doxyfile` and added some detailed description on some of the function groups. A proofread is required, but I hope you can integrate it in the readthedocs documentation since the API was very poorly documented.

straithe commented:
Thank you for this contribution! I'll review this either in the next few days or in January.

martinling reviewed (changes requested):
Hi @Sasszem,

This is a huge amount of work, and I'm extremely impressed that you've taken on such a large task, especially as your first contribution to the HackRF project. Documenting the libhackrf API fully is something that's been on the todo list for a long time.

You have been very thorough in both discovering, and explaining, all sorts of details that were previously only understood by very few people.

I have made a first pass at reviewing this, mainly to clarify some of the more obscure details, to answer some questions that you raised in comments, and suggest some places where the wording might be made clearer or more accurate.

Because of the large amount of work you have done, there is a large number of comments attached to this review, but I hope you appreciate that this does not correspond to any large amount of criticism from me. The vast majority of what you have written is very accurate and I hope that after some adjustments we will be able to get it merged.

I expect that @mossmann will have some further input to make on some of the technical details, and @straithe may want to do a more thorough review than I have of the introductory text which you have added at the top of hackrf.h.

Sasszem commented on host/libhackrf/src/hackrf.h:
```diff
@@ -43,88 +43,288 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
+	 */
 	HACKRF_SUCCESS = 0,
+	/**
+	 * TRUE value, only used by hackrf_is_streaming
```
Since this breaks "conventional" error handling (`!result` or `result != HACKRF_SUCCESS`) I do think this case requires some special attention, mostly at the documentation of the functions that return this specific value, but a mention of this odd-ish behavior is worth mentioning here in my opinion. I agree that a list of functions is not future-proof though, so I guess I'll remove that.

Sasszem commented on host/libhackrf/src/hackrf.h:
```diff
@@ -165,274 +385,816 @@ typedef struct hackrf_device hackrf_device;
+	int buffer_length;
+	/** number of buffer bytes that were transferred */
+	int valid_length;
+	/** RX libusb context. No usage found, possibly not intended for external usage. */
+	void* rx_ctx;
+	/** TX libusb context. No usage found, possibly not intended for external useage. */
+	void* tx_ctx;
```
I had the same idea, but somewhere I read that it's libusb context, so I wrote that.

Sasszem commented on host/libhackrf/src/hackrf.h:
```diff
@@ -165,274 +385,816 @@ typedef struct hackrf_device hackrf_device;
 
-/* external amp, bool on/off */
+/**
+ * Enable/disable 14dB RF amplifier
+ * 
+ * Enable / disable the ~14dB RF RX/TX amplifiers U13/U25 via controlling switches U9 and U14.
```
I was a bit confused there, but the [FAQ section on the docs](https://hackrf.readthedocs.io/en/latest/faq.html#what-gain-controls-are-provided-by-hackrf) says 14dB. Shall I open a separate issue on that or fix it in this PR?

martinling replied:
Yeah, unfortunately that error has crept into a lot of places. I wrote a bit more about it in https://github.com/greatscottgadgets/hackrf/issues/1059#issuecomment-1060038293.

I think it would be best to keep this PR focused on the Doxygen side, and correct other documentation in another PR.

Sasszem replied:
Opened issue #1247

Sasszem commented on host/libhackrf/src/hackrf.h:
```diff
@@ -165,274 +385,816 @@ typedef struct hackrf_device hackrf_device;
+ * 
+ * Enable or disable the **3.3V (max 50mA)** bias-tee (antenna port power). Defaults to disabled.
+ * 
+ * **WARNING:** the firmware auto-disables this after returning to IDLE mode, so a perma-set is not possible, which can cause problems with software not supporting this option.
```
I read this is somewhat a trap for newcomers (see the issues about bias-tee usage), so I think it still deserves a `NOTE`.

Sasszem commented on host/libhackrf/src/hackrf.h:
```diff
@@ -442,74 +1204,252 @@ extern ADDAPI int ADDCALL hackrf_init_sweep(
 	hackrf_device* device,
 	uint8_t* boards);
 
+// TODO: why arent't all functions use adresses?
```
How does that work? I know we have a limit `HACKRF_OPERACAKE_MAX_FREQ_RANGES` (set to 8), so how does specifying more than that work? Will the ranges be distributed across the boards (in groups of 8)? In what order then?

martinling replied:
Sorry, I'd misunderstood - the [docs clarify that](https://hackrf.readthedocs.io/en/latest/opera_cake.html):

> Although multiple Opera Cakes on a single HackRF One may be set to frequency mode at the same time, they share a single switching plan. This can be useful, for example, for a filter bank consisting of eight filters.

and:

> Although multiple Opera Cakes on a single HackRF One may be set to time mode at the same time, they share a single switching plan.

So the reason those functions don't take an address parameter is that the effects apply to all addresses.

Sasszem commented on host/libhackrf/src/hackrf.h:
```diff
@@ -45,54 +45,481 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
+ * 
+ * # Library internals
+ * 
+ * The library uses `libusb` (version 0.1) to communicate with HackRF hardware. It uses both the synchronous and asynchronous API for communication (asynchronous for streaming data to/from the device, and synchronous for everything else). The asynchronous API requires to periodically call a variant of `libusb_handle_events`, so the library creates a new "transfer thread" for each device doing tha ...
```
Sorry, I received false information on that from one of the `libusb` contributors (based on the include).

Sasszem commented:
Thank you for your quick review!

I quickly read through the requested changes and answered those that I was uncertain about, but I shall be able to fix them in the following days.

Sasszem commented on host/libhackrf/src/hackrf.c:
```diff
@@ -287,7 +287,7 @@ static int allocate_transfers(hackrf_device* const device)
 		device->transfers = (struct libusb_transfer**) calloc(
 			TRANSFER_COUNT,
-			sizeof(struct libusb_transfer));
+			sizeof(struct libusb_transfer)); // TODO: this wastes memory (possibly)
```
Opened issue #1246

Sasszem commented on host/libhackrf/src/hackrf.h:
```diff
@@ -165,274 +385,816 @@ typedef struct hackrf_device hackrf_device;
+/**
+ * Start receiving
+ * 
+ * Shuld be called after setting gains
```
In this case, I think the default values of gains should be documented (possibly at the gain setting functions, and at the corresponding overview section)

martinling replied:
There aren't any default gains in the libhackrf API, though. The device retains the last gains that were set.

There are defaults in the firmware, but libhackrf could be talking to any version of the firmware, including any number of forked versions, so I don't think the libhackrf API documentation should be making any guarantees about what the firmware defaults are.

I think it's fine to point out that the caller _may_ want to set the gains before calling `hackrf_start_rx`, I just think the language needs to be quite careful and specific, because "should" can be taken to mean that it's required to do it before every call.

Sasszem commented on host/libhackrf/src/hackrf.h:
```diff
@@ -165,274 +385,816 @@ typedef struct hackrf_device hackrf_device;
+ */
 extern ADDAPI int ADDCALL hackrf_spiflash_erase(hackrf_device* device);
 
+// TODO: is the 256 bytes limit a hard limit? hackrf_spiflash uses it, but not sure if it's necessary
```
I think this function should check this and return an error in this case.

martinling replied:
The function does return an error in this case. A USB device can reject a control request. The firmware rejects the request if the write size is too large. So `libusb_control_request` will return an error, and so in turn will `hackrf_spiflash_write`.

Sasszem commented:
@martinling I fixed most of the issues you pointed out. I still have left 3 todos in:
- Opera Cake addresses
- exact frequency resolution
- UI

martinling reviewed (commented):
The new changes look good. I've flagged up a few typos etc.

Sasszem commented:
General ping, is there any info on my questions?
- frequency resolution
- `hackrf_enable_ui`
I'd also throw in the question of the LED mapping in `hackrf_set_leds`, introduced in 173612ac, as it's unclear. My guess is N-N-N-N-N-TX-RX-USB for the HackRF One, but I don't know the name of the 4th LED on the Rad1o platform, and could not find any information about it.

martinling commented:
The LEDs on rad1o aren't labelled with names or meanings. They're just kinda scattered around the board. From LSB to MSB they go: bottom left green, bottom right green, top left green, top centre red.

The purpose of `hackrf_enable_ui` is basically just to turn on or off the display, for platforms that have one. The standard HackRF firmware includes very basic support for the Portapack add-on, and will use its screen to display a simplified schematic of the RF path, indicating the current mode, frequency, bandwidth and gain settings. On the smaller built-in screen on rad1o, the same information is shown in text form. The UI is read-only, there's no way to interact with it. It is shown by default, but `hackrf_enable_ui` allows disabling it and turning the screen & backlight off.

Frequency resolution is complicated. The RF centre frequency results from a combination of the MAX2837 IF frequency and the RFFC5072 LO frequency, both of which come from PLLs configured with fractional dividers. There isn't any single consistent step value - the resolution varies with where you are in each range, and there are often multiple ways to get close to the same RF centre frequency. The tuning algorithm jumps around between them somewhat in order to dodge internal spurs where possible. Sometimes we tweak the algorithm to get better results for certain frequencies.

Rather than trying to guarantee anything about how close it will get in the API documentation, something that @mossmann and I were discussing recently was adding new API calls to allow reading back and setting exact fractional frequencies for the LO, IF and RF.

Sasszem commented:
Thank you, I will update the docs with this info soon.

Sasszem commented:
Rebased to the current master and updated the docs:
- documented `hackrf_set_leds`
- documented `hackrf_board_id_platform`
- added info on frequency resolution
- added info on `hackrf_set_ui_enable`
- a few other small changes

straithe commented:
Can you please move the Doxyfile to the docs folder under a new folder named "doxygen"?

straithe commented on host/libhackrf/src/Doxyfile:
```diff
@@ -0,0 +1,2658 @@
+# for a project that appears at the top of each page and should give viewer a
+# quick idea about the purpose of the project. Keep the description short.
+
+PROJECT_BRIEF          = "hackrf SDR platform library"
```
Please update the text from "hackrf SDR platform library" to "HackRF SDR platform library".

straithe commented on host/libhackrf/src/Doxyfile:
Please change from "a4" to "letter".

straithe commented on host/libhackrf/src/Doxyfile:
Please update this line from "hackrf.h" to I"../../host/libhackrf/src/hackrf.h" and then move this Doxyfile from "hackrf/host/libhackrf/src/" to a folder named "doxygen" under the "/hackrf/docs/" folder.

straithe commented:
Changes look good. Thank you, @Sasszem.

martinling commented on host/libhackrf/src/hackrf.h:
```diff
@@ -43,88 +43,730 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
+ * 
+ * # Querying device information
+ * 
+ * ## board ID
```
This and other headings in this section should have their first letters capitalized.

martinling commented on host/libhackrf/src/hackrf.h:
````diff
@@ -43,88 +43,730 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
+ *  ```
+ * 
+ * 
+ * ## Partid serial number
+ * 
+ * "part ID number", part id of the MCU. Read via @ref hackrf_board_partid_serialno_read. See the documentation of the MCU for details!
````
```suggestion
 * ## Part ID and serial number
 * 
 * The part ID and serial number of the MCU. Read via @ref hackrf_board_partid_serialno_read. See the documentation of the MCU for details.
```

martinling commented on host/libhackrf/src/hackrf.h:
```diff
@@ -43,88 +43,730 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
+ * 
+ * There are 5 different amplifiers in the HackRF One. Most of them have variable gain, but some of them can be either enabled / disabled. Please note that most of the gain settings are not precise, and they depend on the used frequency as well.
+ * 
+ * ![hackrf components](https://hackrf.readthedocs.io/en/latest/_images/block-diagram.png)
+ * (image taken from https://hackrf.readthedocs.io/en/latest/hardware_components.html)
```
An empty line is needed here to stop text following the large image on a line.

```suggestion
 * ![hackrf components](https://hackrf.readthedocs.io/en/latest/_images/block-diagram.png)
 *
 * (image taken from https://hackrf.readthedocs.io/en/latest/hardware_components.html)
```

martinling commented on host/libhackrf/src/hackrf.h:
```diff
@@ -43,88 +43,730 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
+ * ## MAX2837 2.3 to 2.7 GHz transceiver
+ * This transceiver chip is the RF modulator/demodulator of the HackRF One. This chip sends/receives analoge I/Q samples to/from the MAX5864 ADC/DAC chip. 
+ * 
+ * It's registers are accessible through the functions @ref hackrf_max2837_read and @ref hackrf_max2837_write
```
typo: It's -> Its

(and in a few more places below)

martinling commented on host/libhackrf/src/hackrf.h:
```diff
@@ -136,305 +778,939 @@ enum operacake_ports {
 extern ADDAPI uint32_t ADDCALL hackrf_compute_baseband_filter_bw(
 	const uint32_t bandwidth_hz);
 
 /* All features below require USB API version 0x1002 or higher) */
```
There's a typo in this and several other comments about required USB API versions.

The USB API version required here is `0x0102`, not `0x1002`. In any case, since the comments are for human consumption I think it would be better to say `1.02`. 

These comments are also inconsistent: they're present for several functions that require API 1.02, but there's no comments for other functions that require later API versions. For instance `hackrf_get_m0_state` requires API 1.06. I think we should either ensure required USB API versions are fully listed, or not mention them.

martinling replied:
(I note that this comment, which predates your PR, is similarly inaccurate and is probably the source of the error!)

martinling commented on host/libhackrf/src/hackrf.h:
```diff
@@ -136,305 +778,939 @@ enum operacake_ports {
+ * 
+ * See the documentation on hardware triggering for details
+ * 
+ * Requires USB API version 0x1002 or higher!
```
There's a typo in this and several other comments about required USB API versions.

The USB API version required here is represented as `0x0102`, not `0x1002`. In any case, since the documentation is for human consumption I think it would be better to say `1.02`. 

These comments are also inconsistent: they're present for several functions that require API 1.02, but there's no comments for other functions that require later API versions. For instance `hackrf_get_m0_state` requires API 1.06. I think we should either ensure required USB API versions are fully listed, or not mention them.

Sasszem replied:
I will do a review of USB API versions in a different commit later.

martinling commented on host/libhackrf/src/hackrf.h:
```diff
@@ -136,305 +778,939 @@ enum operacake_ports {
+/**
+ * Set baseband filter bandwidth
+ * 
+ * Possible values: 1.75/2.5/3.5/5/5.5/6/7/8/9/10/12/14/15/20/24/28MHz, default <= 0.75 * sample_rate_hz
```
This list of values is hard to read in this format. 
```suggestion
 * Possible values: 1.75, 2.5, 3.5, 5, 5.5, 6, 7, 8, 9, 10, 12, 14, 15, 20, 24, or 28 MHz.
```
The default could be better rendered as `\f$ \le 0.75 \cdot F_s\f$`, both here and elsewhere.

martinling commented on host/libhackrf/src/hackrf.h:
```diff
@@ -136,305 +778,939 @@ enum operacake_ports {
+/**
+ * Set sample rate
+ * 
+ * Sample rate should be in the range 2-20MHz, with the default being 10MHz. Lower & higher values are technicly possibly, but the performance is not guaranteed.
```
typo: technically possible

martinling commented on host/libhackrf/src/hackrf.h:
```diff
@@ -136,305 +778,939 @@ enum operacake_ports {
 	const uint8_t value);
 
+/**
+ * Read board partid serial number
```
```suggestion
 * Read board part ID and serial number
```

martinling reviewed (changes requested):
Having previously focused my review on the comments added, I've now spent some time looking over the Doxygen output in both HTML and PDF formats.

I've suggested some modifications to the Doxygen settings to improve the output for our needs, and some corrections, typos and formatting suggestions for things I noticed whilst looking through the output.

martinling reviewed (commented):
Thanks for the changes @Sasszem, all looks good.

The only point I'm concerned with now is the documentation of what USB API versions are required for each function. We should either include this information for all functions with version requirements, or leave it out of the current PR, but not leave it incompletely documented.

straithe commented:
@Sasszem are you more comfortable adding in the rest of the information to document functions which require a minimum USB API version? Or would you prefer removing the partial information?

Sasszem commented:
> @Sasszem are you more comfortable adding in the rest of the information to document functions which require a minimum USB API version? Or would you prefer removing the partial information?

I think I'll remove it for now, and open a separate issue for updating it in a separate PR.

martinling reviewed (approved):
All looking great. Thanks @Sasszem for all your work writing up and adjusting this.
